### ISSUE TITLE
Performance not as expected

### ISSUE THREAD

**Issue by ljayx** [NVIDIA/TensorRT-LLM#2179]
Hi experts,

I'm using the latest trtllm code for my llama2-like model inference(smoothquant) which tunned supporting long context(32k) from llama2, but I observed that the performance may not be quite normal, because using vllm(fp16) as an inference engine is a bit better under the same conditions.

You could check below performance table, the 1st token latency is high, and the qps(sequece-level) is not high enough.

The model conversion and building steps listed as below.

Could you help check it? 

**TRTLLM code:** main branch with commit id b8fc6633ba27550dea5f33641f5d94b6c7f02125 (merged @2024/08/27)
**GPU:** A800-SXM4-80GB*4
**docker:** nvidia-cuda:12.4.1-devel-ubuntu22.04
**model:** llama2-like, tunned supportting 32k.

convert:
```bash
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 python3 ./convert_checkpoint.py \
    --model_dir=/llama2-70b-32k/ \
    --output_dir=/llama2-70b-32k/checkpoint-tp4-pp1 \
    --dtype=float16 \
    --tp_size=4 \
    --pp_size=1 \
    --smoothquant 0.5
```

build:
```bash
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 trtllm-build --checkpoint_dir=/llama2-70b-32k/checkpoint-tp4-pp1 \
    --output_dir=/llama2-70b-32k/checkpoint-tp4-pp1/4-gpus-i24k-s32k-t8k-pagedcontxt \
    --max_input_len=24576 \
    --max_seq_len=32768 \
    --max_num_tokens=8192 \
    --gemm_plugin=auto \
    --gpt_attention_plugin=auto \
    --kv_cache_type=paged \
    --use_paged_context_fmha="enable" \
    --remove_input_padding="enable" \
    --context_fmha="enable"
```

performance:
```
┌──────────────┬──────────────────┬────────┬────────┬──────────────┬────────┬─────────────────────────┬──────────┬─────────┬────────┬───────────────────────┬─────────────────┬────────────────────────────────┬───────────────────────────┬─────────────────────────────────────────────┐
│     Job      │      加速卡      │ Engine │ 并发数 │ 输出吞吐/tps │ Stream │   1st token latency/s   │ 总请求数 │ 折算QPS │ 成功率 │        显存/GB        │  加速卡利用率/% │            输入长度            │          输出长度         │                     备注                    │
├──────────────┼──────────────────┼────────┼────────┼──────────────┼────────┼─────────────────────────┼──────────┼─────────┼────────┼───────────────────────┼─────────────────┼────────────────────────────────┼───────────────────────────┼─────────────────────────────────────────────┤
│ 240830-LXhSb │ A800-SXM4-80GB*4 │ TRTLLM │   8    │    101.27    │   Y    │  1.94 (0.30/3.38/3.55)  │    8     │   0.75  │ 100.0% │ 73.8 (73.8/73.8/73.8) │  74 (0/100/100) │ 3451.6 (1506.0/5563.3/6992.0)  │  135.1 (25.0/241.7/290.0) │ TRTLLM0.13.0.dev2024082700/fp16/总耗时10.7s │
│ 240830-LXhSb │ A800-SXM4-80GB*4 │ TRTLLM │   16   │    80.52     │   Y    │  4.06 (0.25/7.81/8.91)  │    16    │   0.53  │ 100.0% │ 73.8 (73.8/73.8/73.8) │  90 (3/100/100) │ 4408.4 (1506.0/6992.0/12167.0) │  152.2 (23.0/411.0/696.0) │ TRTLLM0.13.0.dev2024082700/fp16/总耗时30.2s │
│ 240830-LXhSb │ A800-SXM4-80GB*4 │ TRTLLM │   24   │    74.54     │   Y    │ 5.89 (0.25/11.40/14.04) │    24    │   0.38  │ 100.0% │ 73.8 (73.8/73.8/73.8) │  94 (0/100/100) │ 4581.7 (1506.0/6992.0/12167.0) │ 195.5 (26.0/277.2/1575.0) │ TRTLLM0.13.0.dev2024082700/fp16/总耗时62.9s │
│ 240830-LXhSb │ A800-SXM4-80GB*4 │ TRTLLM │   32   │    89.71     │   Y    │ 7.90 (0.25/14.40/17.42) │    32    │   0.58  │ 100.0% │ 73.8 (73.8/73.8/73.8) │ 99 (99/100/100) │ 4298.8 (1506.0/6992.0/12167.0) │ 154.3 (26.0/250.6/1508.0) │ TRTLLM0.13.0.dev2024082700/fp16/总耗时55.0s │
└──────────────┴──────────────────┴────────┴────────┴──────────────┴────────┴─────────────────────────┴──────────┴─────────┴────────┴───────────────────────┴─────────────────┴────────────────────────────────┴───────────────────────────┴─────────────────────────────────────────────┘
```

vllm performance with same model and dataset, but used fp16, not quantized:

![Image](https://github.com/user-attachments/assets/9bdeeb3e-a016-4fe0-91ce-f7e0f491cf5e)

**Comment by lfr-0531** [comment]
How did you get the performance of TensorRT-LLM? Can you share the cmds that can help us reproduce? 

We recommand using the `gptManagerBenchmark` to test the model performance. Please refer to: https://github.com/NVIDIA/TensorRT-LLM/blob/v0.12.0/benchmarks/cpp/README.md#benchmark-c-runtime

**Comment by ljayx** [comment]
I used hlapi. Service is very similar to the example apps: https://github.com/NVIDIA/TensorRT-LLM/blob/v0.12.0/examples/apps/fastapi_server.py

below is the core code of loading and inference functions.

```python
# loading:
            # FIX: tensorrt_llm/hlapi/llm.py: __init__: add parameter enable_chunked_context
            llm_args = {
                "tokenizer":
                    tokenizer,
                "kv_cache_config":
                    KvCacheConfig(
                        free_gpu_memory_fraction=0.9, enable_block_reuse=False,
                    ),
                "enable_chunked_context": True,
                "scheduler_config":
                    SchedulerConfig(CapacitySchedulerPolicy.GUARANTEED_NO_EVICT)
            }
            self.llm = LLM(self.engine_dir, **llm_args)

# inference:
        sampling_config = SamplingParams(stop_token_ids=self.stop_token_ids, max_new_tokens=max_new_tokens)
        promise = self.llm.generate_async(prompt, sampling_params=sampling_config, streaming=True)
        async for output in promise:
            yield GenerateResponse(
                text=output.outputs[0].text,
                usage=Usage(
                    prompt_tokens=len(output.prompt_token_ids),
                    completion_tokens=len(output.outputs[0].token_ids),
                    total_tokens=len(output.prompt_token_ids) + len(output.outputs[0].token_ids)
                )
            )
```

**Comment by lfr-0531** [comment]
Sorry for the late response. It is a known issue for the bad performance of hlapi with `streaming=True`. We will fix it in v0.13.0 that will be released soon.
